### PR TITLE
Bc 18965

### PR DESCRIPTION
**Pull request by MarvaZy**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,21 +7,21 @@ Yellowbox makes it easy to run docker containers as part of black box tests.
...
 from yellowbox.extras import RedisService
 
 def test_black_box():
-    docker_client = DockerClient.from_env()
-    with RedisService.run(docker_client) as redis:
-        redis_port = redis.client_port()  # this the host port the redis
-        ...  # run your black box test here
-    # yellowbox will automatically close the service when exiting the scope
+    with docker_client() as docker_client:
+      with RedisService.run(docker_client) as redis:
```
Comment (bentheiii) [inline README.md:15]: chain

File: README.md [hunk 2]
```diff
@@ -7,21 +7,21 @@ Yellowbox makes it easy to run docker containers as part of black box tests.
...
+          ...  # run your black box test here
+      # yellowbox will automatically close the service when exiting the scope
 
 def test_black_box_with_initial_data():
     # you can use the service's built-in utility functions to
     # easily interoperate the service
-    docker_client = DockerClient.from_env()
-    with RedisService.run(docker_client) as redis:
+    with docker_client() as docker_client:    
+      with RedisService.run(docker_client) as redis:
```
Comment (bentheiii) [inline README.md:24]: chain

File: README.md [hunk 3]
```diff
@@ -39,38 +39,37 @@ Note: all these extras require additional dependencies as specified in the proje
...
-        # Container/YellowService, retrieves the container's alias(es) on 
-        # the network, and disconnects the two when done
-        another_container = docker_client.containers.create("my-image", 
-            environment={"RABBITMQ_HOSTNAME": alias[0]}
-        )
-        with connect(network, another_container):
-            another_container.start()
-            another_container.wait()
+    with docker_client() as docker_client:
+      with RabbitMQService.run(docker_client) as rabbit, \
```
Comment (bentheiii) [inline README.md:48]: chain

File: tests/conftest.py [hunk 4]
```diff
@@ -12,15 +11,22 @@
...
+        ret = DockerClient.from_env()
+        ret.ping()
+    except Exception:
+        ret = DockerClient(base_url='tcp://localhost:2375')
+        ret.ping()
+
+    try:
+        yield ret
+    finally:
+        ret.close()
```
Comment (bentheiii) [inline tests/conftest.py:24]: preserve the `with closing`

File: tests/test_network.py [hunk 5]
```diff
@@ -33,3 +33,14 @@ def test_connect_shared_network(docker_client: DockerClient, create_and_pull):
                 container.start()
                 return_status = container.wait()
             assert return_status["StatusCode"] == 0
+
+
+def test_connect_with_run(docker_client: DockerClient, create_and_pull):
+    with temp_network(docker_client) as network:
+        with RedisService.run(docker_client, network=network) as service:
```
Comment (bentheiii) [inline tests/test_network.py:40]: chain

File: yellowbox/docker_utils.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+from contextlib import contextmanager
+
+from docker import DockerClient
+
+
+@contextmanager
+def docker_client():
+    """
+    Starts docker client, while being WSL-aware
```
Comment (bentheiii) [inline yellowbox/docker_utils.py:9]:
```suggestion
    Starts docker client from the environment, with a fallback to default HTTP port (for running from within virtual machines)
```

File: yellowbox/networks.py [hunk 7]
```diff
@@ -27,9 +27,9 @@ def temp_network(client: DockerClient, name=None, *args, **kwargs):
     Network will be automatically removed upon context manager completion.
 
     Example:
-        >>> client = DockerClient.from_env()
-        >>> with temp_network(client) as network:
-        ...  print(network.name)
+        >>> with docker_client() as ret:
+        ...  with temp_network(client) as network:
```
Comment (bentheiii) [inline yellowbox/networks.py:31]: chain

File: yellowbox/subclasses.py [hunk 8]
```diff
@@ -115,22 +116,30 @@ def service_name(cls):
...
-        with service:
-            yield service
+        if network:
+            cm = networks_mod.connect(network, service)
+        else:
+            cm = nullcontext()
+
+        with cm as aliases:
+            if aliases:
+                service.alias = aliases[0]
```
Comment (bentheiii) [inline yellowbox/subclasses.py:142]: delete

File: tests/conftest.py [hunk 9]
```diff
@@ -12,15 +11,22 @@
...
 
 
 @fixture
 def host_ip():
-    if platform.system() == "Linux":
+    if platform.system() == "Linux" and ('Microsoft' not in platform.uname().release):
         return '172.17.0.1'
     else:
         return 'host.docker.internal'
```
Comment (bentheiii) [inline tests/conftest.py:32]: delete this whole fixture, use global var instead

File: yellowbox/utils.py [hunk 10]
```diff
@@ -34,7 +34,7 @@ def get_free_port():
         return s.getsockname()[1]
 
 
-if platform.system() == "Linux":
+if platform.system() == "Linux" and ('Microsoft' not in platform.uname().release):
```
Comment (bentheiii) [inline yellowbox/utils.py:37]: document

File: tests/extras/test_azure_storage.py [hunk 11]
```diff
@@ -7,6 +7,7 @@
 from yellowbox.extras.azure_storage import BLOB_STORAGE_DEFAULT_PORT, DEFAULT_ACCOUNT_KEY, DEFAULT_ACCOUNT_NAME, \
     BlobStorageService
 from yellowbox.networks import temp_network, connect
+from yellowbox.utils import docker_host_name as host_ip
```
Comment (bentheiii) [inline tests/extras/test_azure_storage.py:10]: useless as import

File: tests/extras/test_kafka.py [hunk 12]
```diff
@@ -1,7 +1,9 @@
 from pytest import mark
 
+from yellowbox.containers import get_aliases
 from yellowbox.extras.kafka import KafkaService
 from yellowbox.networks import temp_network, connect
+from yellowbox.utils import docker_host_name as host_ip
```
Comment (bentheiii) [inline tests/extras/test_kafka.py:6]: no import as

File: yellowbox/docker_utils.py [hunk 13]
```diff
@@ -1,21 +1,19 @@
...
 def docker_client():
     """
-    Starts docker client, while being WSL-aware
+    Starts docker client from the environment, with a fallback to default HTTP port
+    (for running from within virtual machines)
     """
     try:
         ret = DockerClient.from_env()
         ret.ping()
     except Exception:
```
Comment (bentheiii) [inline yellowbox/docker_utils.py:15]: U2U put you can put up a pragma nocov here

File: yellowbox/subclasses.py [hunk 14]
```diff
@@ -137,9 +147,8 @@ def run(cls: Type[_T], docker_client: DockerClient, *, spinner: bool = True,
         else:
             cm = nullcontext()
```
Comment (bentheiii) [inline yellowbox/subclasses.py:148]: rename variable to something coherent

File: yellowbox/networks.py [hunk 15]
```diff
@@ -27,9 +27,8 @@ def temp_network(client: DockerClient, name=None, *args, **kwargs):
     Network will be automatically removed upon context manager completion.
 
     Example:
-        >>> client = DockerClient.from_env()
-        >>> with temp_network(client) as network:
-        ...  print(network.name)
+        >>> with docker_client() as ret, temp_network(client) as network:
```
Comment (bharel) [inline yellowbox/networks.py:30]: Example broken

File: yellowbox/docker_utils.py [hunk 16]
```diff
@@ -0,0 +1,19 @@
+from contextlib import contextmanager, closing
```
Comment (bharel) [inline yellowbox/docker_utils.py:1]: Rename file.
yellowbox.clients maybe?

**Review by bharel — COMMENTED** [review]
Boop

**Comment by bharel** [comment]
General comment for pull request name, don't use Bc ... It's an open source project, that name is meaningless :-)

**Review by bentheiii — APPROVED** [review]

**Pull request merged**